File: communicationAPI/src/main/java/eu/cybergeiger/communication/MessageHandler.java
package eu.cybergeiger.communication;

import ch.fhnw.geiger.totalcross.ByteArrayInputStream;
import ch.fhnw.geiger.totalcross.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import totalcross.net.Socket;

/**
 * Class to handle incoming messages.
 */
public class MessageHandler implements Runnable {
  private final Socket socket;
  private final LocalApi localApi;

  public MessageHandler(Socket s, LocalApi api) {
    this.socket = s;
    this.localApi = api;
  }

  @Override
  public void run() {
    Message msg;
    try (InputStream in = socket.asInputStream()) {
      // read bytes
      byte[] inputData = new byte[4096];
      int numRead;
      ByteArrayOutputStream buffer = new ByteArrayOutputStream();
      while ((numRead = in.read(inputData, 0, inputData.length)) != -1) {
        // shorten to the written data, TODO maybe not needed?
        byte[] convert = Arrays.copyOfRange(inputData, 0, numRead);
        buffer.write(convert);
      }

      ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(buffer.toByteArray());
      msg = Message.fromByteArray(byteArrayInputStream);

      PluginInformation pluginInformation = new PluginInformation(null, 0);
      System.out.println("## got message ("+msg+")");
      localApi.receivedMessage(pluginInformation, msg);
    } catch (IOException ioe) {
      // TODO handle communications error
      //throw new CommunicationException("Communication Error", ioe);
      ioe.printStackTrace();
    }
  }
}

File: lib/src/communication/MessageHandler.dart
import 'dart:io';

import 'LocalApi.dart';
import 'Message.dart';
import 'PluginInformation.dart';

/// Class to handle incoming messages.
class MessageHandler /*implements Runnable*/ {
  final Socket socket;
  final LocalApi localApi;

  MessageHandler(this.socket, this.localApi);

  void run() {
    Message msg;
    try {
      InputStream in_ = socket.asInputStream();
      // read bytes
      var inputData = List<int>.filled(4096, 0);
      int numRead;
      ByteArrayOutputStream buffer = ByteArrayOutputStream();
      while ((numRead = in_.read(inputData, 0, inputData.length)) != -1) {
        // shorten to the written data, TODO maybe not needed?
        var convert = inputData.sublist(0, numRead);
        buffer.write(convert);
      }

      ByteArrayInputStream byteArrayInputStream =
          ByteArrayInputStream(buffer.toByteArray());
      msg = Message.fromByteArray(byteArrayInputStream);

      var pluginInformation = PluginInformation(null, 0);
      print('## got message (' + msg.toString() + ')');
      localApi.receivedMessage(pluginInformation, msg);
    } on IOException catch (ioe) {
      // TODO handle communications error
      //throw new CommunicationException("Communication Error", ioe);
      print(ioe);
    }
  }
}
